### ISSUE TITLE
user input for authentication

### ISSUE THREAD

**Issue by fawda123** [niklasvm/shiny-strava#6]
Maybe put this as a feature on the app itself rather than in global.R.

So I definitely broke the dev version of this with 86c89849f382d9ec2436f9260953b3dcccdd5c8b but it might be worth pursuing.  I guess it depends on the intended use of the app.  Having it under the hood in global.R would be preferred if 1) the user doesn't want this info displayed on the screen and 2) the user is planning on hosting the app on their own server, i.e., authentication is hard-coded on a deployed app.  Otherwise, I see the app as being something we host somewhere and then anybody can just enter their own info.  

Either way, I broke the app in my latest commit so this may not be worth it.  Feel free to revert if you like, I don't have a strong preference.

**Comment by niklasvm** [comment]
So is the idea that the user can enter their api credentials through some form fields as opposed to setting them in the code itself?

Sounds like a cool idea. I'll take a look at your commit and see how to go about making it workeable

**Comment by niklasvm** [comment]
I've managed to get it working in https://github.com/niklasvm/shiny-strava/commit/be08ef5aeb6ae22497556db947a74d5ecc1b7311. Check it out.

**Comment by niklasvm** [comment]
I've added a flag that the author of the app can set: `ask_api_credentials`. If TRUE it will ask for API credentials, if FALSE it will pull them from the environment variables. This covers both use cases.

**Comment by fawda123** [comment]
Awesome, that looks great.  However, the `textInput` boxes for authentication were capturing the input too quickly.  I tried to include an `actionButton` to submit the info after clicking e38967d7f09262308cc1c01209de49a82739bb20.  This then creates the authentication link button.  It's not ideal  - it would be better to have the `actionButton` and authentication link in the same place so both actions (submitting user credentials and authentication) are completed with one click.  Does that make sense?

**Comment by niklasvm** [comment]
yup. I'll give it a test. I may have had some of the variables loaded in my environment when I tested the code yesterday

**Comment by fawda123** [comment]
Yea I had to remove the cache folder and set the appropriate variables in `global.R` first.

**Comment by niklasvm** [comment]
I think I've managed to get it to work. The link behind the authentication image updates dynamically when you change any of the inputs. Could you test?

**Comment by fawda123** [comment]
Attempting to check on this but see #8

**Comment by fawda123** [comment]
I'm still having some issues with the authentication.  I've changed the global options in `global.R`:

```r
# show application api inputs
ask_api_credentials <- T

# whether to load cached data (must have authenticated before)
cache <- F
```
I then add my client ID, secret, and app URL on the form.   The app fails to authenticate and I get a bad request when I step into the `get_stoken()` server function, looking at `app_parameters$token_data`:

```r
$`message`
[1] "Bad Request"
```
I know my credentials are good because they work when I authenticate using functions in the rStrava package.  

Any idea what's going on?  I tried troubleshooting on my own and I'm wondering if the app secret gets used correctly.  I can't find anywhere in the server code where this gets used during the authentication process.

**Comment by niklasvm** [comment]
I think I've found the issue and fixed it. I've tested it when putting credentials in manually as well as loading them from the `credentials.R` file.

**Comment by fawda123** [comment]
Hm, I'm still having some issues. My token still doesn't seem to be authorizing correctly.  I stepped into the
`server.R` file on line 203 to view the `stoken` object and this is what I get:

```r
Browse[1]> stoken
<request>
Headers:
* Authorization: Bearer 
```

And in case you need this:
```r
Browse[1]> sessionInfo()
R version 3.5.1 (2018-07-02)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                           LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] shinythemes_1.1.1    shinydashboard_0.7.0 leaflet_2.0.1        DT_0.4               glue_1.2.0           jsonlite_1.5        
 [7] httr_1.3.1           lubridate_1.7.4      forcats_0.3.0        stringr_1.3.1        dplyr_0.7.5          purrr_0.2.5         
[13] readr_1.1.1          tidyr_0.8.1          tibble_1.4.2         ggplot2_2.2.1        tidyverse_1.2.1      rStrava_0.16.9000   
[19] shiny_1.1.0         

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.17     lattice_0.20-35  assertthat_0.2.0 digest_0.6.15    packrat_0.4.9-3  psych_1.8.4      mime_0.5        
 [8] R6_2.2.2         cellranger_1.1.0 plyr_1.8.4       pillar_1.2.3     rlang_0.2.1      curl_3.2         lazyeval_0.2.1  
[15] readxl_1.1.0     rstudioapi_0.7   foreign_0.8-70   htmlwidgets_1.2  RCurl_1.95-4.10  munsell_0.5.0    broom_0.4.4     
[22] compiler_3.5.1   httpuv_1.4.3     modelr_0.1.2     pkgconfig_2.0.1  mnormt_1.5-5     htmltools_0.3.6  tidyselect_0.2.4
[29] XML_3.98-1.11    crayon_1.3.4     later_0.7.3      bitops_1.0-6     grid_3.5.1       nlme_3.1-137     xtable_1.8-2    
[36] gtable_0.2.0     magrittr_1.5     scales_0.5.0     cli_1.0.0        stringi_1.1.7    reshape2_1.4.3   promises_1.0.1  
[43] bindrcpp_0.2.2   xml2_1.2.0       tools_3.5.1      crosstalk_1.0.0  hms_0.4.2        parallel_3.5.1   yaml_2.1.18     
[50] colorspace_1.3-2 rvest_0.3.2      bindr_0.1.1      haven_1.1.1     
```

**Comment by niklasvm** [comment]
I’ve added logging that prints to the console just before it attempts to
authenticate. It should print the client id, secret and authorisation code.
Does it print out all 3?

**Comment by niklasvm** [comment]
You can also look at the elements of token_data.  It should contain the access token string

**Comment by fawda123** [comment]
Nope, still not getting anything.  The credentials aren't printed in the console:

```r
> runApp()
No credentials found in environment, displaying api credentials form

Listening on http://127.0.0.1:3533
This weekUsing client id: NA
Using secret: 
```
The token_data isn't showing anything either. 

```r
Browse[2]> app_parameters$token_data
$`message`
[1] "Bad Request"

$errors
$errors[[1]]
$errors[[1]]$`resource`
[1] "Application"

$errors[[1]]$field
[1] "client_id"

$errors[[1]]$code
[1] "invalid"
```

**Comment by niklasvm** [comment]
I'm going to add some more detailed logging to the app to try workout why its not working.

I've tested on windows and linux on R 3.5.1 and the authentication worked

**Comment by fawda123** [comment]
Ok, thanks!  I know this is a pain to diagnose since I'm the one having the issue.  I really want to get this working though!

**Comment by niklasvm** [comment]
No worries.

**Comment by niklasvm** [comment]
I have added some logging which I hope will identify the authentication issue.
On a side note I'm also testing the googlePolylines decode functionality.

My packrat is giving some issues so you'l need to manually install:
```{r}
install.packages(c('googlePolylines','logging'))
```

**Comment by fawda123** [comment]
Awesome, thanks Nick.  I just tried it and the logging is working just fine but again still having authentication issues.  The first three lines are saved in the log before I try to authenticate:

```
2018-08-30 08:18:45 INFO:authentication:No credentials found in environment, displaying api credentials form
2018-08-30 08:18:45 INFO:authentication:Set strava app url to http://127.0.0.1:1234
2018-08-30 08:18:45 INFO:authentication:Load file depencencies
```

Then when I entered my credentials and click the connect to Strava button, I get this line:
```
2018-08-30 08:19:44 ERROR:authentication:strava_app_secret is blank
```

It's like the credentials I enter never get passed to where they need to go on the server side.  It's weird, I can step into the server script on line 112 with the browser and I can see the credentials getting added to the `app_parameters` object as I enter them on the app.  I can also see them being added to the `authorisation_url`.  This is what I get after I've entered the client id, secret, and app URL: https://www.strava.com/oauth/authorize?client_id=3067&response_type=code&redirect_uri=http://127.0.0.1:3706&approval_prompt=auto&state=

Then when I click the authorise button again on the app and step back into the server code, the credentials are gone from the authorisation URL: https://www.strava.com/oauth/authorize?client_id=NA&response_type=code&redirect_uri=http://127.0.0.1:1234&approval_prompt=auto&state=

Then of course the app fails to authenticate when that empty URL gets sent to a web browser.  

The log says that the app secret is blank, which might give us a clue because I can't for the life of me figure out where this is used during the authentication process. Is this supposed to be part of the authorisation URL or is it used elsewhere?

**Comment by niklasvm** [comment]
Ok, I think we're making progress. It seems like the credentials aren't persisted properly. I think this has to do with with application scope. Its entirely possible that when you routed back to the app url on from the strava authentication page that it starts a new application session. I'm going to set it to write the captured credentials directly to disk. I think it also makes sense to split it into two parts:

1. App authentication
2. User athentication

Will give you a heads up when I've made the changes.

**Comment by fawda123** [comment]
Alright, sounds good.  I should have also mentioned that when the app fails to authenticate after trying to launch from the first browser, a new session is opened in my web browser.  Not sure if that has to do with the credential persistence issue but thought it might be informative.

**Comment by niklasvm** [comment]
I've pushed some updates which I think should solve the authentication problem. The api credentials are written to disk after being captured so no need to worry about starting a new session.

I've also fixed my packrat issues so all new packages will install automatically.

**Comment by fawda123** [comment]
Alright, this is fantastic! I was able to authenticate and download my activities.  I added shinyjs to `global.R` and also modified `tidy_activities()` in 719017077c928dfb65b1ec9fb3f50a9c73d77b4b.  I don't have heart rate data so I had to add a check in the function to get the data to compile correctly.  

This app is really slick and I appreciate the time you've spent on this so far.  I'll get you some comments on general layout and some other ideas in #7.
